torch/nested/_internal/nested_tensor.py

```diff
@@ -3,23 +3,224 @@
 import torch
 from torch._C import DispatchKey, DispatchKeySet
 from torch._prims_common import is_expandable_to
-from torch.fx.experimental.symbolic_shapes import has_free_symbols
 from torch.utils.weak import WeakTensorKeyDictionary
 from typing import *  # noqa: F403
+import weakref
 
-_tensor_id_counter = 0
-_tensor_symint_registry = WeakTensorKeyDictionary()
+def _get_nested_int(equiv_set, vec):
+    return torch._C._get_nested_int(equiv_set, coeff=1, vec=vec)
 
+class NestedIntRegistry():
+    # Class to manage the association between 1d tensors (vec) and nested ints.
+    # Use this class to (1) obtain nested int from vec (2) read/write metadata
+    # associated with vecs/nested ints
+    #
+    # Pass a vec to maybe_create to create a new nested int. Multiple nested int
+    # can be associated with a vec. vec are then grouped into equivalence sets
+    # so that comparisons between the vec's nested ints can be made by comparing
+    # their equiv set. Morally two vec are in the same equiv set if and only if
+    # they have the same data, e.g., vec.cpu() and vec.cuda() belong to the same
```

Comment:
This doesn't seem to match what we discussed in composability sync. @cpuhrsch's proposal, which we accepted in composability, was that morally, the size tensor is canonically a **cpu tensor**, which may have a cached cuda tensor attached to it (which we use to avoid HtoD transfer when we're running kernels.) If CUDA let us transmit as many integer parameters in CUDA launch, we would have done that instead.

In this regime, it doesn't make sense to me talk about two vecs of differing devices being in the same set. I'm never creating a nested int with a cuda tensor. I only ever have nested ints of cpu tensors (and then mumble mumble cuda caching). This has implications on public facing NT API: you can't create a NT from a CUDA tensor sizes *only* (a DtoH sync must happen if you do), but *this makes sense*, because how are you going to check if this has the same size as something else if you don't do this sync!

Reply:
> You can't create a NT from a CUDA tensor sizes only (a DtoH sync must happen if you do)

Ah I might have missed the discussion on the trade offs with this implication in the composability meeting, so after realizing that this was a new implication, I ended up convincing myself to go the other way. Here's what I understand the trade off to be:

Pros of "morally a cpu tensor"
- Avoid the implicit DtoH sync when doing `zeros([j0])
Cons:
- Adding a sync for when we create NT if offsets passed is CUDA. 

I think the cons outweigh the pros:
- This DtoH sync is very common in the use cases we're working with (some kind of torch rec JaggedTensor with lengths on CUDA) and will continue to be common until NT is used everywhere.
- The NT(offsets_on_cuda) DtoH sync isn't actually completely explicit - sure it makes sense under the  "morally a cpu tensor" framework, but users won't know that unless they study NT docs
- Implicit DtoH sync is rare - it only occurs when someone does calls a factory function where they intend for the result to be a cpu tensor.
- Implicit DtoH sync case can be mitigated by adding a warning or something.

Reply:
> Adding a sync for when we create NT if offsets passed is CUDA

Well, if this is the tall pole in the tent, one could imagine lazily populating the CPU tensor in the CUDA -> CPU case, in case you never end up actually needing to consult it. I think it is at least worth thinking about this more, because there is a big pro: size equality follows intuitive rules: two tensors compare equal if their nested sizes logically compare equal. No "well, this CPU nested size is never equal to a CUDA size", no concept of device for nested size.

By the way, while JT "apparently" doesn't hold its sizes on CPU, it will happily do tons of DtoH syncs on some things on its API:

```
    def to_dense(self) -> List[torch.Tensor]:
        tensor_list = []
        for index in range(self.offsets().size(0) - 1):
            offset = self.offsets()[index].item()
            next_offset = self.offsets()[index + 1].item()
            tensor_list.append(self.values()[offset:next_offset])
        return tensor_list
```

KJT is even more interesting: we can think of it as a NT with two dimensions before the jagged dimension (feature, batch). The sizes on the batch dimension are not cached on CPU (same as NT). But the sizes on feature dimension are!

```
        values: torch.Tensor = [V0, V1, V2, V3, V4, V5, V6, V7]  # V == any tensor datatype
        weights: torch.Tensor = [W0, W1, W2, W3, W4, W5, W6, W7]  # W == any tensor datatype
        lengths: torch.Tensor = [2, 0, 1, 1, 1, 3]  # representing the jagged slice
        offsets: torch.Tensor = [0, 2, 2, 3, 4, 5, 8]  # offsets from 0 for each jagged slice
        keys: List[str] = ["Feature0", "Feature1"]  # correspond to each value of dim_0
        index_per_key: Dict[str, int] = {"Feature0": 0, "Feature1": 1}  # index for each key
        offset_per_key: List[int] = [0, 3, 8]  # start offset for each key and final offset
```

So one cannot even say torchrec isn't doing this, it is, when it is extremely helpful.

Reply:
> Well, if this is the tall pole in the tent, one could imagine lazily populating the CPU tensor in the CUDA -> CPU case, in case you never end up actually needing to consult it. I think it is at least worth thinking about this more, because there is a big pro: size equality follows intuitive rules: two tensors compare equal if their nested sizes logically compare equal. No "well, this CPU nested size is never equal to a CUDA size", no concept of device for nested size.

Ahh this is definitely my fault for procrastinating actually fleshing out the device handling code lol, but if I am understanding correctly, in the current impl, we DO obey those rules around shape equality (since vec_cpu and vec_cuda are in the same equiv set), and the intention IS actually to do the lazily population (because we want to avoid doing .to() to populate that cpu tensor immediately, and we still need some tensor placeholder to put in the weak tensor key to equiv set map, our only choice is to declare the cuda tensor as canonical). 

To be clear, all canonical means here is: all alive nested ints objects in that equiv set must point to the same vec, and that vec we call the canonical vec. The idea that a canonical vec exists at all is intended to entirely be an implementation detail. It should not matter to users whether that canonical vec is a cpu vec or a cuda vec. Concretely, if I constructed a NJT with cuda offsets, that cuda offsets would become the canonical vec purely because it is the first vec for bookkeeping reasons, but morally, the nested int returned is still cpu. That is, if I do `nt_zeros = zeros([j0])` that would create a cpu tensor. In fact, because every equivalence set also has a single canonical nested int, nt_zeros (on cpu) would hold the same exact nested int object (which still references the canonical vec which is on cuda) as the original cuda NT the nested int was extracted from.

Anyway, initially I had imagined our concern to be: if we're saying that size is device agnostic, then it might be unexpected if torch.zeros(nt_cuda.shape) incurred a implicit DtoH sync. But now if we're okay with saying: "we still have a morally a cpu tensor and the reason we incurred a DtoH sync there is that the cpu offsets is materialized lazily", that sounds a lot like a good argument to stick with the design in this PR today!

Reply:
So this is where... you need an actual PR description... that says what the updated design is lol.

Based on your description here, I still have more questions:

* If two CUDA sizes tensors happen to be equal, but we have to actually perform an equality test to find out this is the case, if I test equality between these two vecs, do you discover that this is the case and merge their equivalence classes?
* What is the ownership relationship between the canonical vec (tensor) and all the other vec (tensors) in the equivalence class? Notably, if I repeatedly cast a cpu vec to cuda, do I get the same cuda vec or different cuda vecs

You don't need to implement the code to figure it out. Spend some time thinking about it and writing out an English spec.

Reply:
> So this is where... you need an actual PR description... that says what the updated design is lol.

Totally agree :P.

> If two CUDA sizes tensors happen to be equal, but we have to actually perform an equality test to find out this is the case, if I test equality between these two vecs, do you discover that this is the case and merge their equivalence classes?

I don't plan to implement merging sets in this PR, but that's certainly a possibility. In the near term, the user is expected to tell us what equiv set a vec belongs to upon registeration via `maybe_create(equiv_set_from=)`. APIs like .to() will automatically use that API. 

> What is the ownership relationship between the canonical vec (tensor) and all the other vec (tensors) in the equivalence class? Notably, if I repeatedly cast a cpu vec to cuda, do I get the same cuda vec or different cuda vecs

Same cuda vec, the equiv set dict will have dedicated "slots": vec_cuda and vec_cpu  that are intended to store cached cuda or cpu copies of the vec. If someone repeatedly casts a nt to a device, we would materialize the offset in the slot of the associated the first time, and return the same tensor the next time. Ownership-wise as long as one vec in the equiv_set is alive, the equiv set's associated meta cache will stay alive. 

This diagram might be helpful:
```
#  symint       vec         equiv_set       cached metadata
# --------------------------------------------------------------------------
#  2*j0 -----\                        /---- weak canonical vec0, weak canonical j0_0
#             v                      | /--> id: int = 0
#  j0_0 ------> vec0 -----> equiv_set0 ---> sum_vec: Optional[SymInt]
#  j0_1 ------^         /-^            \--> vec_{cuda,cpu}: Optional[Tensor]
#                      /
#               vec1--/               /---- weak vec2, weak j1_0
#                                    | /--> id: int = 1
#  j1_0 ------> vec2 -----> equiv_set1 ---> ...
```